clmm/utils.py

```diff
@@ -863,3 +871,76 @@ def _srd_z_distrib(redshift, is_cdf=False):
         return redshift0**(alpha+1)*gammainc((alpha+1)/beta, (redshift/redshift0)**beta)/beta*gamma((alpha+1)/beta)
     else:
         return (redshift**alpha)*np.exp(-(redshift/redshift0)**beta)
+
+def _draw_random_points_from_distribution(xmin, xmax, nobj, dist_func, xstep=0.001):
+    """Draw random points with a given distribution.
+
+    Uses a sampling technique found in Numerical Recipes in C, Chap 7.2: Transformation Method.
+
+    Parameters
+    ----------
+    xmin : float
+        The minimum source redshift allowed.
+    xmax : float, optional
+        If source redshifts are drawn, the maximum source redshift
+    nobj : float
+        Number of galaxies to generate
+    dist_func : function
+        Function of the required distribution
+    xstep : float
+        Size of the step to interpolate the culmulative distribution.
+
+    Returns
+    -------
+    ndarray
+        Random points with dist_func distribution
+    """
+    steps = int((xmax-xmin)/xstep)+1
+    xdomain = np.linspace(xmin, xmax, steps)
+    # Cumulative probability function of the redshift distribution
+    #probdist = np.vectorize(lambda zmax: integrate.quad(dist_func, xmin, zmax)[0])(xdomain)
+    probdist = dist_func(xdomain, is_cdf=True)-dist_func(xmin, is_cdf=True)
+    # Get random values for probdist
+    uniform_deviate = np.random.uniform(probdist.min(), probdist.max(), nobj)
+    return interp1d(probdist, xdomain, kind='linear')(uniform_deviate)
+
+def _draw_random_points_from_tab_distribution(x_tab, pdf_tab, nobj=1, xmin=None, xmax=None):
+    """Draw random points from a tabulated distribution.
+
+    Parameters
+    ----------
+    x_tab : array-like
+        Values for which the tabulated pdf is provided
+    pdf_tab : array-like
+        Value of the pdf at the x_tab locations
+    nobj : int, optional
+        Number of random samples to generate. Default is 1.
+    xmin : float
+        Lower bound to draw redshift. Default is the min(x_tab)
+    xmax : float
+        Upper bound to draw redshift. Default is the max(x_tab)
+
+    Returns
+    -------
+    samples : ndarray
+        Random points following the pdf_tab distribution
+    """
+    x_tab = np.array(x_tab)
+    pdf_tab = np.array(pdf_tab)
+    # Compute the tabulated cumulative distribution
+    if xmin:
+        pdf_tab = pdf_tab[x_tab>=xmin]
```

Comment:
these cuts here are throwing data away. They are the responsible for the empty gap at 0.3<x<0.325 in [this](https://github.com/LSSTDESC/CLMM/pull/526#issuecomment-1178487483) fig. I think we sould:

- check if `x_tab.min()<xmin`
  - if False, raise a warning with the value of `x_tab.min()`
  - else, we compute the cdf with the full pdf but only sample for cdf(xsample)>cdf(`xmin`)
  
  same thing for `xmax`. You can do it by just computing the cdf of these values:
  ```
  cdf_xmin, cdf_xmax = 0.0, 1.0
  if xmin or xmax:
      cdf_interp = interp1d(x_tab, cdf, kind='linear')
      if xmin:
          cdf_xmin = cdf_interp(xmin)
      if xmax:
          cdf_xmax = cdf_interp(xmax)
  samples = inv_cdf(np.random.random(nobj)*(cdf_xmax-cdf_xmin)+cdf_xmin)
  ```